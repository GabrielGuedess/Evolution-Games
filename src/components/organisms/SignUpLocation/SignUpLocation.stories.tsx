import { Story, Meta } from '@storybook/react/types-6-0';

import { SignUpLocation, SignUpLocationProps } from './SignUpLocation';

const locationInputs = {
  cep: '',
  logradouro: '',
  numero: '',
  complemento: '',
  referencia: '',
  bairro: '',
  cidade: '',
  uf: '',
};

const setLocationInputs = () => ({});
const setCurrentStep = () => ({});

export default {
  title: 'Organisms/SignUpLocation',
  component: SignUpLocation,
  args: {
    locationInputs,
    setLocationInputs,
    setCurrentStep,
    currentStep: 3,
  },
} as Meta<SignUpLocationProps>;

export const Default: Story<SignUpLocationProps> = args => (
  <SignUpLocation {...args} />
);
