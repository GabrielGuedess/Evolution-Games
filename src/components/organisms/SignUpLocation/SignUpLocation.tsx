import { useState, Dispatch, SetStateAction } from 'react';
import ReactInputMask from 'react-input-mask';

import { CircleNotch } from 'phosphor-react';
import { FieldErrors, signUpLocationValidate } from 'utils/validations';

import Button from 'components/atoms/Button/Button';
import { InputForm } from 'components/atoms/InputForm/InputForm';

import * as S from './SignUpLocation.styles';

export type LocationInputs = {
  cep: string;
  logradouro: string;
  numero: string;
  complemento: string;
  referencia: string;
  bairro: string;
  cidade: string;
  uf: string;
};

export type SignUpLocationProps = {
  locationInputs: LocationInputs;
  setLocationInputs: Dispatch<SetStateAction<LocationInputs>>;
  currentStep: number;
  setCurrentStep: (step: number) => void;
};

type FetchCEP = {
  cep: string;
  logradouro: string;
  complemento: string;
  bairro: string;
  localidade: string;
  uf: string;
  ibge: string;
  gia: string;
  ddd: string;
  siafi: string;
  erro?: boolean;
};

export const SignUpLocation = ({
  locationInputs,
  setLocationInputs,
  currentStep,
  setCurrentStep,
}: SignUpLocationProps) => {
  const [errorMessages, setErrorMessages] = useState<FieldErrors>({});
  const [isLoading, setIsLoading] = useState(false);

  const getCEP = async () => {
    try {
      const errors = signUpLocationValidate({
        ...locationInputs,
        cep: locationInputs.cep,
      });

      setErrorMessages({ ...errors });

      if (!errorMessages.cep) {
        setIsLoading(true);

        const data: FetchCEP = await (
          await fetch(`https://viacep.com.br/ws/${locationInputs.cep}/json/`)
        ).json();

        if (data.erro) {
          setErrorMessages({ cep: 'CEP não encontrado' });
          setIsLoading(false);

          setLocationInputs({
            uf: '',
            cidade: '',
            bairro: '',
            logradouro: '',
            complemento: '',
            cep: '',
            referencia: '',
            numero: '',
          });

          return;
        }

        setErrorMessages({});

        setLocationInputs({
          uf: data.uf,
          cidade: data.localidade,
          bairro: data.bairro,
          logradouro: data.logradouro,
          complemento: data.complemento,
          cep: data.cep,
          referencia: '',
          numero: '',
        });

        setIsLoading(false);
      }
    } catch (error) {
      setIsLoading(false);
    }
  };

  const handleInput = (field: keyof LocationInputs, value: string) => {
    setLocationInputs(s => ({ ...s, [field]: value }));

    const errors = signUpLocationValidate({
      ...locationInputs,
      [field]: value,
    });

    setErrorMessages({ ...errorMessages, [field]: errors[field] });
  };

  const handleSubmit = async (event: React.FormEvent) => {
    event.preventDefault();

    if (
      Object.values(errorMessages).some(item => item !== undefined) ||
      Object.values(locationInputs).some(item => item === '')
    ) {
      setErrorMessages(errorMessages);
      return;
    }

    setErrorMessages({});
    setCurrentStep(currentStep + 1);
  };

  return (
    <S.Wrapper role="form" onSubmit={handleSubmit}>
      <S.InputBox>
        <InputForm
          required
          as={ReactInputMask}
          mask="99999-999"
          type="text"
          name="cep"
          label="CEP"
          labelFor="cep"
          value={locationInputs.cep}
          placeholder="Insira seu cep"
          errorMessage={errorMessages?.cep}
          onChange={e =>
            handleInput('cep', e.target.value.replace(/[^0-9]/g, ''))
          }
          isInvalid={!!errorMessages?.cep}
          icon={
            isLoading ? (
              <CircleNotch size={20} className="animation-spinner">
                <animateTransform
                  attributeName="transform"
                  attributeType="XML"
                  type="rotate"
                  dur="0.6s"
                  from="0 0 0"
                  to="360 0 0"
                  repeatCount="indefinite"
                />
              </CircleNotch>
            ) : (
              <S.Search size={20} onClick={getCEP} />
            )
          }
        />

        <InputForm
          required
          type="text"
          name="logradouro"
          label="Logradouro"
          labelFor="logradouro"
          value={locationInputs.logradouro}
          placeholder="Insira seu logradouro"
          errorMessage={errorMessages?.logradouro}
          onChange={e => handleInput('logradouro', e.target.value)}
          isInvalid={!!errorMessages?.logradouro}
        />

        <InputForm
          required
          type="number"
          name="numero"
          label="Número"
          labelFor="numero"
          value={locationInputs.numero}
          placeholder="Insira seu número"
          errorMessage={errorMessages?.numero}
          onChange={e => handleInput('numero', e.target.value)}
          isInvalid={!!errorMessages?.numero}
        />

        <InputForm
          required
          type="text"
          name="complemento"
          label="Complemento"
          labelFor="complemento"
          value={locationInputs.complemento}
          placeholder="Insira seu complemento"
          errorMessage={errorMessages?.complemento}
          onChange={e => handleInput('complemento', e.target.value)}
          isInvalid={!!errorMessages?.complemento}
        />

        <InputForm
          required
          type="text"
          name="referencia"
          label="Referência"
          labelFor="referencia"
          value={locationInputs.referencia}
          placeholder="Insira sua referencia"
          errorMessage={errorMessages?.referencia}
          onChange={e => handleInput('referencia', e.target.value)}
          isInvalid={!!errorMessages?.referencia}
        />

        <InputForm
          required
          type="text"
          name="bairro"
          label="Bairro"
          labelFor="bairro"
          value={locationInputs.bairro}
          placeholder="Insira seu bairro"
          errorMessage={errorMessages?.bairro}
          onChange={e => handleInput('bairro', e.target.value)}
          isInvalid={!!errorMessages?.bairro}
        />

        <InputForm
          required
          type="text"
          name="cidade"
          label="Cidade"
          labelFor="cidade"
          value={locationInputs.cidade}
          placeholder="Insira sua cidade"
          errorMessage={errorMessages?.cidade}
          onChange={e => handleInput('cidade', e.target.value)}
          isInvalid={!!errorMessages?.cidade}
        />

        <InputForm
          required
          type="text"
          name="uf"
          label="UF"
          labelFor="uf"
          value={locationInputs.uf}
          placeholder="Insira sua UF"
          errorMessage={errorMessages?.uf}
          onChange={e => handleInput('uf', e.target.value)}
          isInvalid={!!errorMessages?.uf}
        />
      </S.InputBox>

      <Button
        disabled={
          Object.values(errorMessages).some(item => item !== undefined) ||
          Object.values(locationInputs).some(item => item === '')
        }
        as="button"
        type="submit"
        size="large"
        fullWidth
      >
        Próximo
      </Button>
    </S.Wrapper>
  );
};
