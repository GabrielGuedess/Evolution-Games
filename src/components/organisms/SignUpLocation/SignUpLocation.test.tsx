import { renderWithProviders } from 'utils/tests/helpers';

import { screen } from '@testing-library/react';

import { SignUpLocation } from './SignUpLocation';

const locationInputs = {
  cep: '',
  logradouro: '',
  numero: '',
  complemento: '',
  referencia: '',
  bairro: '',
  cidade: '',
  uf: '',
};

const setLocationInputs = jest.fn();
const setCurrentStep = jest.fn();

describe('<SignUpLocation />', () => {
  it('should render the heading', () => {
    const { container } = renderWithProviders(
      <SignUpLocation
        locationInputs={locationInputs}
        setLocationInputs={setLocationInputs}
        currentStep={0}
        setCurrentStep={setCurrentStep}
      />,
    );

    expect(screen.getByRole('form')).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });
});
