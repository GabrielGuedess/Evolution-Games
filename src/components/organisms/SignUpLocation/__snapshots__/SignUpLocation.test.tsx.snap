// Jest Snapshot v1, https://goo.gl/fbAQLP

exports['<SignUpLocation /> should render the heading 1'] = `
.c9 {
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  cursor: pointer;
  position: relative;
  outline: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  border-radius: 5px;
  -webkit-transition: all 0.3s cubic-bezier(0.68,-0.6,0.32,1.6);
  transition: all 0.3s cubic-bezier(0.68,-0.6,0.32,1.6);
  font-size: 1.6rem;
  padding: 0.8rem 4.8rem;
  background: #7C3AED;
  color: #F4F4F5;
  width: 100%;
}

.c9:not(:disabled):is(:hover,:focus) {
  box-shadow: 0 0.8rem 2.8rem rgba(118,74,241,0.5);
  color: #F4F4F5;
}

.c9 .effect {
  fill: #7C3AED;
}

.c9:disabled {
  cursor: not-allowed;
  -webkit-filter: saturate(30%);
  filter: saturate(30%);
}

.c9:not(:disabled):is(:hover,:focus) {
  -webkit-filter: brightness(1.1);
  filter: brightness(1.1);
}

.c9:not(:disabled):active {
  -webkit-filter: brightness(0.9);
  filter: brightness(0.9);
}

.c9 span {
  font-size: 1.6rem;
  font-weight: 500;
  line-height: 1;
  display: block;
  position: relative;
  -webkit-transition: -webkit-transform 0.3s cubic-bezier(0.68,-0.6,0.32,1.6);
  -webkit-transition: transform 0.3s cubic-bezier(0.68,-0.6,0.32,1.6);
  transition: transform 0.3s cubic-bezier(0.68,-0.6,0.32,1.6);
  z-index: 1;
}

.c9:not(:disabled):is(:hover,:focus) span {
  -webkit-transform: scale(1.05);
  -ms-transform: scale(1.05);
  transform: scale(1.05);
}

.c9:not(:disabled):is(:hover,:focus) .icon {
  -webkit-transform: scale(1.2);
  -ms-transform: scale(1.2);
  transform: scale(1.2);
}

.c9:not(:disabled):active span {
  -webkit-transform: scale(0.95);
  -ms-transform: scale(0.95);
  transform: scale(0.95);
}

.c9 svg {
  -webkit-transition: -webkit-transform 0.3s cubic-bezier(0.68,-0.6,0.32,1.6);
  -webkit-transition: transform 0.3s cubic-bezier(0.68,-0.6,0.32,1.6);
  transition: transform 0.3s cubic-bezier(0.68,-0.6,0.32,1.6);
}

.c9 .effect {
  position: absolute;
  top: initial;
  left: initial;
  width: 100%;
  height: 100%;
  z-index: -1;
}

.c9 .effect path {
  -webkit-transition: 0.3s cubic-bezier(0.68,-0.6,0.32,1.6);
  transition: 0.3s cubic-bezier(0.68,-0.6,0.32,1.6);
}

.c9:not(:disabled):is(:hover,:focus) .effect path {
  d: path( 'M0,0 C0,-5 100,-5 100,0 C105,0 105,100 100,100 C100,105 0,105 0,100 C-5,100 -5,0 0,0' );
}

.c9:not(:disabled):active .effect path {
  d: path( 'M0 5C0 2.23858 2.23858 0 5 0H95C97.7614 0 100 2.23858 100 5V95C100 97.7614 97.7614 100 95 100H5C2.23858 100 0 97.7614 0 95V5Z' );
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c3 {
  position: relative;
}

.c6 {
  position: absolute;
  left: 1rem;
  top: 1.7rem;
  padding: 0 3px;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
  z-index: 2;
}

.c6::before {
  content: '';
  position: absolute;
  inset: 0;
  width: 100%;
  z-index: -1;
}

.c4 {
  width: 100%;
  font-size: 1.8rem;
  font-weight: 400;
  font-style: normal;
  line-height: 3.2rem;
  border-width: 2px;
  border-style: solid;
  border-color: #272530;
  background-color: transparent;
  border-radius: 0.5rem;
  padding: 0.8rem;
  color: #FFFFFF;
  outline: none;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
}

.c4::-webkit-calendar-picker-indicator {
  background-image: url('data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" width="16" height="15" viewBox="0 0 24 24"><path fill="%23272530" d="M20 3h-1V1h-2v2H7V1H5v2H4c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm0 18H4V8h16v13z"/></svg>');
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
}

.c4 .c5 {
  color: #FFFFFF;
}

.c4:-webkit-autofill {
  -webkit-box-shadow: 0 0 0px 100rem #04040C inset;
  -webkit-text-fill-color: #FFFFFF;
}

.c4:-webkit-autofill ~ div,
.c4:-webkit-autofill:hover ~ div,
.c4:-webkit-autofill:focus ~ div,
.c4:-webkit-autofill:active ~ div {
  color: #717B8A;
}

.c4::-webkit-input-placeholder {
  font-size: 1.4rem;
  opacity: 0;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
}

.c4::-moz-placeholder {
  font-size: 1.4rem;
  opacity: 0;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
}

.c4:-ms-input-placeholder {
  font-size: 1.4rem;
  opacity: 0;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
}

.c4::placeholder {
  font-size: 1.4rem;
  opacity: 0;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
}

.c4:focus {
  border-color: #7C3AED;
}

.c4:focus::-webkit-calendar-picker-indicator {
  background-image: url('data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" width="16" height="15" viewBox="0 0 24 24"><path fill="%237C3AED" d="M20 3h-1V1h-2v2H7V1H5v2H4c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm0 18H4V8h16v13z"/></svg>');
}

.c4:focus::-webkit-input-placeholder {
  opacity: 1;
}

.c4:focus::-moz-placeholder {
  opacity: 1;
}

.c4:focus:-ms-input-placeholder {
  opacity: 1;
}

.c4:focus::placeholder {
  opacity: 1;
}

.c4:is(:focus,:valid) + .c5 {
  top: -2rem;
  color: #7C3AED;
  font-size: 1.2rem;
}

.c4::selection {
  color: #FFFFFF;
  background: #7C3AED;
}

.c7 {
  width: 50px;
  height: 50px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: absolute;
  right: 0;
  top: 0;
}

.c0 {
  width: 100%;
  max-width: 40.4rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 3.2rem;
}

.c0 > button span {
  padding: 0.9rem 0;
}

.c1 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 3.2rem;
}

.c8 {
  cursor: pointer;
  color: #717B8A;
}

.c8:is(:hover,:focus) {
  color: #7C3AED;
}

@media (min-width:450px) {
  .c9 .effect {
    top: -5%;
    left: -5%;
    width: 110%;
    height: 110%;
  }
}

@media (min-width:768px) {
  .c6::before {
    background: #04040C;
  }
}

@media (min-width:768px) {
  .c4:is(:focus,:valid) + .c5 {
    top: -8px;
  }
}

@media (min-width:768px) {
  .c1 {
    gap: 1.6rem;
  }
}

<form
  class="c0"
  role="form"
>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <input
          autocomplete="on"
          class="c4"
          id="cep"
          name="cep"
          placeholder="Insira seu cep"
          required=""
          type="text"
          value=""
        />
        <label
          aria-label="cep"
          class="c5 c6"
          for="cep"
        >
          CEP
        </label>
        <div
          class="c7"
        >
          <svg
            class="c8"
            fill="currentColor"
            height="20"
            viewBox="0 0 256 256"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              fill="none"
              height="256"
              width="256"
            />
            <circle
              cx="116"
              cy="116"
              fill="none"
              r="84"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="16"
            />
            <line
              fill="none"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="16"
              x1="175.4"
              x2="224"
              y1="175.4"
              y2="224"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <input
          autocomplete="on"
          class="c4"
          id="logradouro"
          name="logradouro"
          placeholder="Insira seu logradouro"
          required=""
          type="text"
          value=""
        />
        <label
          aria-label="logradouro"
          class="c5 c6"
          for="logradouro"
        >
          Logradouro
        </label>
      </div>
    </div>
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <input
          autocomplete="on"
          class="c4"
          id="numero"
          name="numero"
          placeholder="Insira seu número"
          required=""
          type="number"
          value=""
        />
        <label
          aria-label="numero"
          class="c5 c6"
          for="numero"
        >
          Número
        </label>
      </div>
    </div>
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <input
          autocomplete="on"
          class="c4"
          id="complemento"
          name="complemento"
          placeholder="Insira seu complemento"
          required=""
          type="text"
          value=""
        />
        <label
          aria-label="complemento"
          class="c5 c6"
          for="complemento"
        >
          Complemento
        </label>
      </div>
    </div>
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <input
          autocomplete="on"
          class="c4"
          id="referencia"
          name="referencia"
          placeholder="Insira sua referencia"
          required=""
          type="text"
          value=""
        />
        <label
          aria-label="referencia"
          class="c5 c6"
          for="referencia"
        >
          Referência
        </label>
      </div>
    </div>
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <input
          autocomplete="on"
          class="c4"
          id="bairro"
          name="bairro"
          placeholder="Insira seu bairro"
          required=""
          type="text"
          value=""
        />
        <label
          aria-label="bairro"
          class="c5 c6"
          for="bairro"
        >
          Bairro
        </label>
      </div>
    </div>
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <input
          autocomplete="on"
          class="c4"
          id="cidade"
          name="cidade"
          placeholder="Insira sua cidade"
          required=""
          type="text"
          value=""
        />
        <label
          aria-label="cidade"
          class="c5 c6"
          for="cidade"
        >
          Cidade
        </label>
      </div>
    </div>
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <input
          autocomplete="on"
          class="c4"
          id="uf"
          name="uf"
          placeholder="Insira sua UF"
          required=""
          type="text"
          value=""
        />
        <label
          aria-label="uf"
          class="c5 c6"
          for="uf"
        >
          UF
        </label>
      </div>
    </div>
  </div>
  <button
    aria-label="Button Action"
    class="c9"
    disabled=""
    type="submit"
  >
    <span>
      Próximo
    </span>
    <svg
      aria-hidden="true"
      class="effect"
      preserveAspectRatio="none"
      viewBox="-5 -5 110 110"
    >
      <path
        d="M0 8C0 3.58172 3.58172 0 8 0H92C96.4183 0 100 3.58172 100 8V92C100 96.4183 96.4183 100 92 100H8C3.58172 100 0 96.4183 0 92V8Z"
      />
    </svg>
  </button>
</form>
`;
